hack/util.sh

```diff
@@ -13,90 +12,116 @@
+  if [[ "${go_version[2]}" == "devel" ]]; then
+    return 0
+  fi
+  util::vercomp "${go_version[2]#go}" "${MIN_Go_VERSION}"
+  if [[ $? -eq 2 ]]; then
     echo "Detected go version: ${go_version[*]}."
-    echo "requires ${MIN_Go_VERSION} or greater."
+    echo "Requires ${MIN_Go_VERSION} or greater."
     echo "Please install ${MIN_Go_VERSION} or later."
     exit 1
   fi
 }
 
-# util::install_helm will install the helm command
+# Version comparison function.
+function util::vercomp {
+  if [[ $1 == $2 ]]; then
+    return 0
+  fi
+  local IFS=.
+  local i ver1=($1) ver2=($2)
+  for ((i=${#ver1[@]}; i<${#ver2[@]}; i++)); do
+    ver1[i]=0
+  done
+  for ((i=0; i<${#ver1[@]}; i++)); do
+    if [[ -z ${ver2[i]} ]]; then
+      ver2[i]=0
+    fi
+    if ((10#${ver1[i]} > 10#${ver2[i]})); then
+      return 1
+    fi
+    if ((10#${ver1[i]} < 10#${ver2[i]})); then
+      return 2
+    fi
+  done
+  return 0
+}
+
+# Install Helm.
 function util::install_helm {
-    curl https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 | bash
+  if util::cmd_exist helm; then
+    echo "Helm is already installed."
+    return 0
+  fi
+  echo "Installing Helm..."
+  curl https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 | bash
 }
 
-# util::exec_cmd will using eval to parse command
-function util::exec_cmd() {
-   if [ $# -eq 0 ] ; then
-     echo "[Error] no command specified for util::exec_cmd()..."
-     exit 2
-   fi
-   local tmpLog=$(mktemp)
-   set +e
-   eval "$@" &> $tmpLog
-   if [ $? -ne 0 ];then
-      echo "[Error] Failed to do $1. detail logs as below:"
-      set +x
-      echo "$(cat $tmpLog)"
-      set -x
-      rm -f $tmpLog
-      exit 3
-   fi
-   echo "$1 successful."
-   rm -f $tmpLog
-   set -e
+# Execute a command and capture output.
+function util::exec_cmd {
+  if [ $# -eq 0 ]; then
+    echo "[Error] No command specified for util::exec_cmd()..."
+    exit 2
+  fi
+  local tmpLog=$(mktemp)
+  set +e
+  "$@" &> "$tmpLog"
+  local ret=$?
+  if [ $ret -ne 0 ]; then
+    echo "[Error] Failed to execute command: $*"
+    echo "Detail logs:"
+    cat "$tmpLog"
+    rm -f "$tmpLog"
+    exit $ret
+  fi
+  echo "Command executed successfully: $*"
+  rm -f "$tmpLog"
+  set -e
 }
 
-### Wait a node reachable
-function util::wait_ip_reachable(){
-    local vm_ip=${1:-""}
-    local loop_time=${2:-"10"}
-    local sleep_time=${2:-"60"}
-    echo "Wait vm_ip=$1 reachable ... "
-    for ((i=1;i<=$((loop_time));i++)); do
-      pingOK=0
-      ping -w 2 -c 1 "${vm_ip}"|grep "0%" || pingOK=false
-      echo "==> ping ""${vm_ip}" $pingOK
-      if [[ ${pingOK} == false ]];then
-        sleep "$sleep_time"
-      else
-        break
-      fi
-      if [ $i -eq $((loop_time)) ];then
-        echo "node not reachable exit!"
-        exit 1
-      fi
-    done
+# Check if an IP is reachable.
```

Comment:
As I said before, the current PR should only focus on solving the serious issue of E2E failure, and format optimization can be separated out.